# The Business context of the data

Our consultancy firm is contracted by a credit card client. The client have offered us a data set with data from account holders. Each row corresponds to one customer's account data. The data contains monthly records, over a period of 6 months. Data from 30,000 acount holders are provided. The data are classified according to whether an account owner has defaulted, after a six month period. In practice this means that the account holder did not make the minimum required payment. 

# The goal for data analytics

The client wants us to develop a model to predict whether in the month after the six-month period of the historical data, e account holder will default or not. 


# The data set

The data set is a modified version of the following: 
https://archive.ics.uci.edu/ml/datasets/default+of+credit+card+clients. 

This is taken from the UCI Machine Learning Repository, a public repository of bencmarking datasets. .
Have a look at that web page to find out more abot the data characteristics. Can you summarise the key characteristics?

Next, let's turn our attention to the code needed to load the data. 

In [110]:
import pandas as pd

Depending on your set up you may have to set the path or not. If you do, make sure it matches your own environment. 
With the first command we imported the pandas data management library. 

Next we will use it to load our first data set. 

This is taken from the UCI Machine Learning Repository [http://archive.ics.uci.edu/ml]. Irvine, CA: University of California, School of Information and Computer Science.


Data loading is done in the next code cell. You may have to adjust the file path. 
Make sure that excel file reading is enabled. 
You can install the relevant library at the command prompt: 
pip install xlrd

In [111]:
#df = pd.read_excel("default_of_credit_card_clients.xls")

In [112]:
df = pd.read_csv('default_of_credit_card_clients.csv',index_col='ID')

This presupposes that we have a directory Data and that the above excel or CSV file is placed there. The file is available on Brightspace. 
Depending on the Python version, the excel or csv reading file might lead to different ways in handling the record IDs. We will use the CSV file but will be handling the IDs differently: we may wish to read them as they are, IDs. But when we need to check for duplicate IDs, we will read the IDs as attributes, so that we can manipulate them. 

We start our exploratory analysis by inspecting the size of the data file

In [134]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE         DOB  \
0  b15e78fe-a223     500000    1          1         1   42  1900-01-01   
1  689ae71c-3f4b      30000    1          2         1   36  1900-01-01   
2  3dc1d96e-35c2      50000    1          1         1    0  1981-10-21   
3  f786754a-eb0b      50000    1          3         1    0  1982-10-22   
4  e6a3c5a6-7647     200000    2          1         2   36  1900-01-01   

           PAY_1  PAY_2  PAY_3  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  \
0             -2     -1      0  ...       6420      49240       6772   
1              1     -1     -1  ...       1170        780          0   
2              0      0      0  ...      31240      20343      12349   
3  Not available     -1     -1  ...       1961       1261       2681   
4             -2     -2     -2  ...          0          0          0   

   PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0     31305     43864      6453     49470      3723    103686   
1       780         0      1170         0         0         0   
2      2017      3000      2000      3007      1003      1500   
3     13196      2866      1961      1261      2681      1261   
4         0         0         0         0         0         0   

   default payment next month  
0                           0  
1                           0  
2                           0  
3                           0  
4                           1  

[5 rows x 26 columns]

In [135]:
#df.to_csv('default_of_credit_card_clients.csv',index=False)

In [136]:
df = pd.read_csv('default_of_credit_card_clients.csv',index_col='ID') # Use 'ID' column as index

In [137]:
df.head()

Unnamed: 0  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE PAY_1  \
ID                                                                          
798fc410-45c1           0      20000    2          2         1   24     2   
8a8c8f3b-8eb4           1     120000    2          2         2   26    -1   
85698822-43f5           2      90000    2          2         2   34     0   
0737c11b-be42           3      50000    2          2         1   37     0   
3b7f77cc-dbc0           4      50000    1          2         1   57    -1   

               PAY_2  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  \
ID                                  ...                                    
798fc410-45c1      2     -1     -1  ...          0          0          0   
8a8c8f3b-8eb4      2      0      0  ...       3272       3455       3261   
85698822-43f5      0      0      0  ...      14331      14948      15549   
0737c11b-be42      0      0      0  ...      28314      28959      29547   
3b7f77cc-dbc0      0     -1      0  ...      20940      19146      19131   

               PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
ID                                                                          
798fc410-45c1         0       689         0         0         0         0   
8a8c8f3b-8eb4         0      1000      1000      1000         0      2000   
85698822-43f5      1518      1500      1000      1000      1000      5000   
0737c11b-be42      2000      2019      1200      1100      1069      1000   
3b7f77cc-dbc0      2000     36681     10000      9000       689       679   

               default payment next month  
ID                                         
798fc410-45c1                           1  
8a8c8f3b-8eb4                           1  
85698822-43f5                           0  
0737c11b-be42                           0  
3b7f77cc-dbc0                           0  

[5 rows x 25 columns]

In [138]:
#df = pd.read_excel("mixed_data.xls")

In [139]:
#df.to_csv('mixed_data.csv',index=False)

In [140]:
df = pd.read_csv('mixed_data.csv',index_col='ID') # Use 'ID' column as index

In [120]:
df.head()

LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE         DOB  \
ID                                                                    
b15e78fe-a223     500000    1          1         1   42  1900-01-01   
689ae71c-3f4b      30000    1          2         1   36  1900-01-01   
3dc1d96e-35c2      50000    1          1         1    0  1981-10-21   
f786754a-eb0b      50000    1          3         1    0  1982-10-22   
e6a3c5a6-7647     200000    2          1         2   36  1900-01-01   

                       PAY_1  PAY_2  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  \
ID                                                 ...                         
b15e78fe-a223             -2     -1      0      0  ...       6420      49240   
689ae71c-3f4b              1     -1     -1     -1  ...       1170        780   
3dc1d96e-35c2              0      0      0      0  ...      31240      20343   
f786754a-eb0b  Not available     -1     -1     -1  ...       1961       1261   
e6a3c5a6-7647             -2     -2     -2     -2  ...          0          0   

               BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  \
ID                                                                           
b15e78fe-a223       6772     31305     43864      6453     49470      3723   
689ae71c-3f4b          0       780         0      1170         0         0   
3dc1d96e-35c2      12349      2017      3000      2000      3007      1003   
f786754a-eb0b       2681     13196      2866      1961      1261      2681   
e6a3c5a6-7647          0         0         0         0         0         0   

               PAY_AMT6  default payment next month  
ID                                                   
b15e78fe-a223    103686                           0  
689ae71c-3f4b         0                           0  
3dc1d96e-35c2      1500                           0  
f786754a-eb0b      1261                           0  
e6a3c5a6-7647         0                           1  

[5 rows x 25 columns]

# Starting with Jupyter notebooks

There are two kinds of cells: markdown and 'code' cells. This one is marked as markdown. The next one is a `markdown` cell, followed by a 'code' cell.

In [121]:
df = pd.read_csv('mixed_data.csv',index_col='ID') # Use 'ID' column as index

In [122]:
df.shape

(30000, 25)

So this file has 30000 records and 25 fields (columns)

We can also execute direct value assignment commands and view their outcomes, as below

In [123]:
df.head()

LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE         DOB  \
ID                                                                    
b15e78fe-a223     500000    1          1         1   42  1900-01-01   
689ae71c-3f4b      30000    1          2         1   36  1900-01-01   
3dc1d96e-35c2      50000    1          1         1    0  1981-10-21   
f786754a-eb0b      50000    1          3         1    0  1982-10-22   
e6a3c5a6-7647     200000    2          1         2   36  1900-01-01   

                       PAY_1  PAY_2  PAY_3  PAY_4  ...  BILL_AMT4  BILL_AMT5  \
ID                                                 ...                         
b15e78fe-a223             -2     -1      0      0  ...       6420      49240   
689ae71c-3f4b              1     -1     -1     -1  ...       1170        780   
3dc1d96e-35c2              0      0      0      0  ...      31240      20343   
f786754a-eb0b  Not available     -1     -1     -1  ...       1961       1261   
e6a3c5a6-7647             -2     -2     -2     -2  ...          0          0   

               BILL_AMT6  PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  \
ID                                                                           
b15e78fe-a223       6772     31305     43864      6453     49470      3723   
689ae71c-3f4b          0       780         0      1170         0         0   
3dc1d96e-35c2      12349      2017      3000      2000      3007      1003   
f786754a-eb0b       2681     13196      2866      1961      1261      2681   
e6a3c5a6-7647          0         0         0         0         0         0   

               PAY_AMT6  default payment next month  
ID                                                   
b15e78fe-a223    103686                           0  
689ae71c-3f4b         0                           0  
3dc1d96e-35c2      1500                           0  
f786754a-eb0b      1261                           0  
e6a3c5a6-7647         0                           1  

[5 rows x 25 columns]

# Step 3: Check Data Integrity

The data set contains monthly credit card account data, for a period of six months. Let's perform a quality check to ensure we have the data for the accounts as expected. The account ID distinguishes one account from the other. We can check unique IDs with Pandas with the function `.nunique()`. But first let's check our data structure. To do so we first build an Index of he different columns in the table. We will use the .columns method of the pandas DataFrame to see the column names. 

In [127]:
df = pd.read_csv('mixed_data.csv') # Do not use 'ID' column as index so as to use it for checking duplicate ID records

In [128]:
df.columns

Index(['ID', 'LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'DOB',
       'PAY_1', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1',
       'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6',
       'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6',
       'default payment next month'],
      dtype='object')

We then check the column headings

The first column is the account ID. The other columns contain the dataset 'fetures': 

LIMIT_BAL: Amount of credit provided (in New Taiwanese (NT) dollar) including individual consumer credit and the family (supplementary) credit.

SEX: Gender (1 = male; 2 = female).

Note

The SEC attribute in this form will not be used and we will not be using gender to determine credit-worthiness (ethical use of data). 

EDUCATION: Education (1 = graduate school; 2 = university; 3 = high school; 4 = others).

MARRIAGE: Marital status (1 = married; 2 = single; 3 = others).

AGE: Age (year).

PAY_1–PAY_6: A record of past payments. Past monthly payments, recorded from April to September, are stored in these columns.

PAY_1 represents the repayment status in September; PAY_2 is the repayment status in August; and so on up to PAY_6, which represents the repayment status in April.

The measurement scale for the repayment status is as follows: -1 = pay duly; 1 = payment delay for 1 month; 2 = payment delay for 2 months; and so on up to 8 = payment delay for 8 months; 9 = payment delay for 9 months and above.

BILL_AMT1–BILL_AMT6: Bill statement amount (in NT dollar).

BILL_AMT1 represents the bill statement amount in September; BILL_AMT2 represents the bill statement amount in August; and so on up to BILL_AMT6, which represents the bill statement amount in April.

PAY_AMT1–PAY_AMT6: Amount of previous payment (NT dollar). PAY_AMT1 represents the amount paid in September; PAY_AMT2 represents the amount paid in August; and so on up to PAY_AMT6, which represents the amount paid in April.
    
We can take a look at the first 5 data rows by invoking the 'head()' method of the data frame. 

In [129]:
df.head()

ID  LIMIT_BAL  SEX  EDUCATION  MARRIAGE  AGE         DOB  \
0  b15e78fe-a223     500000    1          1         1   42  1900-01-01   
1  689ae71c-3f4b      30000    1          2         1   36  1900-01-01   
2  3dc1d96e-35c2      50000    1          1         1    0  1981-10-21   
3  f786754a-eb0b      50000    1          3         1    0  1982-10-22   
4  e6a3c5a6-7647     200000    2          1         2   36  1900-01-01   

           PAY_1  PAY_2  PAY_3  ...  BILL_AMT4  BILL_AMT5  BILL_AMT6  \
0             -2     -1      0  ...       6420      49240       6772   
1              1     -1     -1  ...       1170        780          0   
2              0      0      0  ...      31240      20343      12349   
3  Not available     -1     -1  ...       1961       1261       2681   
4             -2     -2     -2  ...          0          0          0   

   PAY_AMT1  PAY_AMT2  PAY_AMT3  PAY_AMT4  PAY_AMT5  PAY_AMT6  \
0     31305     43864      6453     49470      3723    103686   
1       780         0      1170         0         0         0   
2      2017      3000      2000      3007      1003      1500   
3     13196      2866      1961      1261      2681      1261   
4         0         0         0         0         0         0   

   default payment next month  
0                           0  
1                           0  
2                           0  
3                           0  
4                           1  

[5 rows x 26 columns]

In [130]:
df.shape

(30000, 26)

We have now read the 'ID' not as ID but as attribute, so that we can check for duplicate IDs. Earlier, we saw the type of data that the ID column contains. It appears to be a unique identifier. But is it? Let's now check the unique IDs

In [131]:
df['ID'].nunique()

29687

That's interesting! We have fewer unique IDs than rows, so we clearly have duplicates! But we don't really know how many IDs are repeated how many times. We can check the number of occurences of each ID by counting them, as follows: 

In [132]:
id_counts = df['ID'].value_counts()
id_counts.head()

4fcef74d-0b3d    2
a7e199e1-2c44    2
a9ce8636-9a88    2
d9dafe9c-b497    2
3367c92b-b1ca    2
Name: ID, dtype: int64

This wasn't too helpful. It only returned the occurences of the first five rows. 

In [133]:
id_counts.value_counts()

1    29374
2      313
Name: ID, dtype: int64

So now we now that we have 29374 IDs which appear only once but we also have 313 which appear twice!